core/state_connector.go

```diff
@@ -14,219 +15,196 @@ import (
+	WithBlockContext(bc vm.BlockContext)
+	BlockContext() vm.BlockContext
 }
 
-func GetTestingChain(chainID *big.Int) bool {
-	return chainID.Cmp(flareChainID) != 0 && chainID.Cmp(songbirdChainID) != 0
+// stateConnector is responsible for calling state connector smart contract based on votes from attestors
+type stateConnector struct {
+	caller Caller
+	msg    Message
 }
 
-func GetStateConnectorActivated(chainID *big.Int, blockTime *big.Int) bool {
-	if GetTestingChain(chainID) {
-		return true
-	} else if chainID.Cmp(flareChainID) == 0 {
-		return blockTime.Cmp(flareStateConnectorActivationTime) >= 0
-	} else if chainID.Cmp(songbirdChainID) == 0 {
-		return blockTime.Cmp(songbirdStateConnectorActivationTime) >= 0
+func newConnector(caller Caller, msg Message) *stateConnector {
+	s := &stateConnector{
+		caller: caller,
+		msg:    msg,
 	}
-	return false
+	return s
 }
 
-func GetStateConnectorContract(chainID *big.Int, blockTime *big.Int) common.Address {
-	switch {
-	case GetStateConnectorActivated(chainID, blockTime) && chainID.Cmp(songbirdChainID) == 0:
-		return common.HexToAddress("0x6b5DEa84F71052c1302b5fe652e17FD442D126a9")
-	default:
-		return common.HexToAddress("0x1000000000000000000000000000000000000001")
-	}
+type attestationVotes struct {
+	reachedMajority    bool
+	majorityDecision   string
+	majorityAttestors  []common.Address
+	divergentAttestors []common.Address
+	abstainedAttestors []common.Address
 }
 
-func GetStateConnectorCoinbaseSignalAddr(chainID *big.Int, blockTime *big.Int) common.Address {
-	switch {
-	default:
-		return common.HexToAddress("0x000000000000000000000000000000000000dEaD")
+func (c *stateConnector) finalizePreviousRound(chainID, timestamp *big.Int, currentRoundNumber []byte) error {
+	instructions := append(attestationSelector(chainID, timestamp), currentRoundNumber[:]...)
+	defaultAttestors, err := c.defaultAttestors(chainID, timestamp)
+	if err != nil {
+		return err
 	}
-}
-
-func SubmitAttestationSelector(chainID *big.Int, blockTime *big.Int) []byte {
-	switch {
-	default:
-		return []byte{0xcf, 0xd1, 0xfd, 0xad}
+	defaultAttestationVotes := c.countAttestations(defaultAttestors, instructions)
+	localAttestors := localAttestors()
+	var finalityReached bool
+	if len(localAttestors) > 0 {
+		localAttestationVotes := c.countAttestations(localAttestors, instructions)
+		if defaultAttestationVotes.reachedMajority && localAttestationVotes.reachedMajority && defaultAttestationVotes.majorityDecision == localAttestationVotes.majorityDecision {
+			finalityReached = true
+		} else if defaultAttestationVotes.reachedMajority && defaultAttestationVotes.majorityDecision != localAttestationVotes.majorityDecision {
+			// FIXME Make a back-up of the current state database, because this node is about to branch from the default set
+		}
+	} else if defaultAttestationVotes.reachedMajority {
+		finalityReached = true
```

Comment:
You could put this in a function to make the code more clear, and just check for the finalityReached, WDYT?